api/src/main/java/org/apache/iceberg/actions/ComputePartitionStats.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.actions;
+
+import org.apache.iceberg.PartitionStatisticsFile;
+
+/** An action that computes and writes the partition statistics of an Iceberg table. */
+public interface ComputePartitionStats
+    extends Action<ComputePartitionStats, ComputePartitionStats.Result> {
+  /**
+   * Choose the table branch to compute stats, if not specified, uses the default branch.
+   *
+   * @param branch branch name of the table for which stats need to be computed
+   * @return this for method chaining
+   */
+  ComputePartitionStats branch(String branch);
```

Comment:
would it better if we expose snapshotId like we do in computeStats (https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/actions/ComputeTableStats.java#L39)

Reply:
The base API was designed per branch initially. So, it is accordingly. 
https://github.com/apache/iceberg/blob/be9808fb3d75b1697fc4ed45c02602de9cfb6371/data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java#L135

Users can create a branch for mapping snapshot, if at all user wants to compute snapshot for specific snapshot.

Reply:
Thanks for sharing the background discussion during puffin stats addition.

Yeah, there is no standard in Iceberg whether to expose branch or snapshot id. 
Snapshot id seems more convenient, I have raised a PR for base interface to adopt snapshot id. 
https://github.com/apache/iceberg/pull/12464

I will rebase this PR after that.   

> or may be support both 

Supporting both is not a good idea as branch is just an alias for snapshot and exposing both will confuse the user and need extra validations if user configures both my mistake.

Reply:
done